File: samples/activation/MarkdownViewer/Helpers/Singleton.cs
// ******************************************************************
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THE CODE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH
// THE CODE OR THE USE OR OTHER DEALINGS IN THE CODE.
// ******************************************************************

using System;
using System.Collections.Concurrent;

namespace MarkdownViewer.Helpers
{
    internal static class Singleton<T> where T : new()
    {
        private static ConcurrentDictionary<Type, T> _instances = new ConcurrentDictionary<Type, T>();

        public static T Instance
        {
            get
            {
                return _instances.GetOrAdd(typeof(T), (t) => new T());
            }
        }
    }
}

File: templates/Uwp/Projects/DefaultVB/Helpers/Singleton.vb
Imports System.Collections.Concurrent

Namespace Helpers
    Friend NotInheritable Class Singleton(Of T As New)
        Private Sub New()
        End Sub

        Private Shared ReadOnly _instances As New ConcurrentDictionary(Of Type, T)()

        Public Shared ReadOnly Property Instance As T
            Get
                Return _instances.GetOrAdd(GetType(T), Function(t) New T())
            End Get
        End Property
    End Class
End Namespace
